src/repository.py

```diff
@@ -48,6 +96,92 @@ def __init__(self, repository: Repo, github_repository: Repository) -> None:
         self._github_repo = github_repository
         self._configure_git_user()
 
+    @cached_property
+    def base_path(self) -> Path:
+        return Path(self._git_repo.working_dir)
+
+    @property
+    def metadata(self) -> Metadata:
+        return get_metadata(self.base_path)
+
+    @property
+    def has_docs_directory(self) -> bool:
+        return has_docs_directory(self.base_path)
+
+    @property
+    def current_branch(self) -> str:
+        return self._git_repo.active_branch.name
+
+    @contextmanager
+    def with_branch(self, branch_name: str) -> 'Client':
+        current_branch = self.current_branch
+
+        yield self.switch(branch_name)
+
+        self.switch(current_branch)
+
+    @property
+    def summary(self) -> DiffSummary:
+        self._git_repo.git.add(DOCUMENTATION_FOLDER_NAME)
+
+        return DiffSummary.from_raw_diff(self._git_repo.index.diff(None)) + \
+            DiffSummary.from_raw_diff(self._git_repo.head.commit.diff())
+
+    def pull(self, branch_name: Optional[str] = None) -> None:
+        if branch_name is None:
+            self._git_repo.git.pull()
+        else:
+            with self.with_branch(branch_name) as repo:
+                repo.pull()
+
+    def switch(self, branch_name: str) -> 'Client':
+        is_dirty = self.is_dirty()
+
+        if is_dirty:
+            self._git_repo.git.add(DOCUMENTATION_FOLDER_NAME)
+            self._git_repo.git.stash()
+
+        try:
+            self._git_repo.git.checkout(branch_name, "--")
+        finally:
+            if is_dirty:
+                try:
+                    self._git_repo.git.stash("pop")
+                except GitCommandError as exc:
+                    if "CONFLICT" in exc.stdout:
+                        self._git_repo.git.checkout("--theirs", DOCUMENTATION_FOLDER_NAME)
```

Comment:
This nesting is getting a bit deep?

Reply:
Yes, I agree. I don't like it either. I was meant to refactor this after the first draft